sklearn/cluster/_k_means.pyx

```diff
@@ -8,167 +8,207 @@
 #
 # License: BSD 3 clause
 
-from libc.math cimport sqrt
 import numpy as np
-import scipy.sparse as sp
 cimport numpy as np
 cimport cython
 from cython cimport floating
+from libc.math cimport sqrt
+
+
+np.import_array()
 
-from sklearn.utils.sparsefuncs_fast import assign_rows_csr
 
 ctypedef np.float64_t DOUBLE
 ctypedef np.int32_t INT
 
-cdef extern from "cblas.h":
-    double ddot "cblas_ddot"(int N, double *X, int incX, double *Y, int incY)
-    float sdot "cblas_sdot"(int N, float *X, int incX, float *Y, int incY)
 
-np.import_array()
+cpdef floating _inertia_dense(np.ndarray[floating, ndim=2, mode='c'] X,
+                              floating[::1] sample_weight, 
+                              floating[:, ::1] centers,
+                              int[::1] labels):
+    """Compute inertia for dense input data
+    
+    Sum of squared distance between each sample and it's assigned center.
+    """
+    cdef:
+        int n_samples = X.shape[0]
+        int n_features = X.shape[1]
+        int i, j, k
+        floating tmp, sample_inertia
+
+        floating inertia = 0.0
+
+    for i in range(n_samples):
```

Comment:
Could this be a `prange`? Cython should be able to properly reduce the sum

You might have to be careful about `sample_inertia` which should be private for this to work I think

Reply:
Sure. It's not the critical part since it's only called once at the end of the run, but faster is better :)